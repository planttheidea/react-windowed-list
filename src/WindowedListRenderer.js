// external dependencies
import PropTypes from 'prop-types';
import React from 'react';
import {createComponent} from 'react-parm';
import {measure} from 'remeasure';

// constants
import {
  HAS_NEW_LIFECYCLE_METHODS,
  OUTER_CONTAINER_STYLE,
  VALID_AXES,
  VALID_AXIS_VALUES,
  VALID_TYPES,
  VALID_TYPE_VALUES
} from './constants';

// instance methods
import {
  getItemSizeAndItemsPerRow,
  getScrollOffset,
  getScrollParent,
  getSizeOfListItem,
  getSpaceBefore,
  getStartAndEnd,
  getVisibleRange,
  renderItems,
  scrollAround,
  scrollTo,
  setReconcileFrameAfterUpdate,
  setScroll,
  setStateIfAppropriate,
  updateFrame,
  updateScrollParent,
  updateSimpleFrame,
  updateUniformFrame,
  updateVariableFrame
} from './instanceMethods';

// lifecycle methods
import {
  onConstruct,
  componentDidMount,
  getDerivedStateFromProps,
  componentWillReceiveProps,
  getSnapshotBeforeUpdate,
  componentDidUpdate,
  componentWillUnmount
} from './lifecycleMethods';

// utils
import {DefaultItemRenderer, DefaultContainerRenderer, getInnerContainerStyle, getListContainerStyle} from './utils';

export const DEFAULT_PARM_OPTIONS = {
  // state
  state: {
    from: 0,
    itemsPerRow: 0,
    size: 0
  },
  // lifecycle methods
  onConstruct,
  componentDidMount,
  componentDidUpdate,
  componentWillUnmount,
  // instance values
  cache: {},
  outerContainer: null,
  reconcileFrameAfterUpdate: null,
  unstableTimeoutId: null,
  updateCounter: 0,
  updateCounterTimeoutId: null,
  // instance methods
  getItemSizeAndItemsPerRow,
  getScrollOffset,
  getScrollParent,
  getSizeOfListItem,
  getSpaceBefore,
  getStartAndEnd,
  getVisibleRange,
  renderItems,
  scrollAround,
  scrollTo,
  setReconcileFrameAfterUpdate,
  setScroll,
  setStateIfAppropriate,
  updateFrame,
  updateScrollParent,
  updateSimpleFrame,
  updateUniformFrame,
  updateVariableFrame
};

export const PARM_OPTIONS = HAS_NEW_LIFECYCLE_METHODS
  ? {
    ...DEFAULT_PARM_OPTIONS,
    getSnapshotBeforeUpdate
  }
  : {
    ...DEFAULT_PARM_OPTIONS,
    componentWillReceiveProps
  };

export const WindowedListRenderer = createComponent(
  ({axis, length, type, usePosition, useTranslate3d}, {getSpaceBefore, renderItems, state: {from, itemsPerRow}}) =>
    type === VALID_TYPES.SIMPLE ? (
      renderItems()
    ) : (
      <div style={OUTER_CONTAINER_STYLE}>
        <div style={getInnerContainerStyle(axis, length, itemsPerRow, getSpaceBefore)}>
          <div style={getListContainerStyle(axis, usePosition, useTranslate3d, from, getSpaceBefore)}>
            {renderItems()}
          </div>
        </div>
      </div>
    ),
  PARM_OPTIONS
);

WindowedListRenderer.displayName = 'WindowedListRenderer';

WindowedListRenderer.propTypes = {
  axis: PropTypes.oneOf(VALID_AXIS_VALUES).isRequired,
  containerRenderer: PropTypes.func.isRequired,
  debounceReconciler: PropTypes.number,
  getEstimatedItemSize: PropTypes.func,
  getItemSize: PropTypes.func,
  getScrollParent: PropTypes.func,
  initialIndex: PropTypes.number.isRequired,
  isHidden: PropTypes.bool.isRequired,
  isLazy: PropTypes.bool.isRequired,
  itemRenderer: PropTypes.func.isRequired,
  length: PropTypes.number.isRequired,
  pageSize: PropTypes.number.isRequired,
  threshold: PropTypes.number.isRequired,
  type: PropTypes.oneOf(VALID_TYPE_VALUES).isRequired,
  usePosition: PropTypes.bool.isRequired,
  useStaticSize: PropTypes.bool.isRequired,
  useTranslate3d: PropTypes.bool.isRequired
};

WindowedListRenderer.defaultProps = {
  axis: VALID_AXES.Y,
  containerRenderer: DefaultContainerRenderer,
  initialIndex: 0,
  isHidden: false,
  isLazy: false,
  itemRenderer: DefaultItemRenderer,
  length: 0,
  minSize: 1,
  pageSize: 10,
  threshold: 100,
  type: VALID_TYPES.SIMPLE,
  usePosition: false,
  useStaticSize: false,
  useTranslate3d: false
};

if (HAS_NEW_LIFECYCLE_METHODS) {
  WindowedListRenderer.getDerivedStateFromProps = getDerivedStateFromProps;
}

export default measure(['height', 'width'], {namespace: '__windowedListMeasurements'})(WindowedListRenderer);
