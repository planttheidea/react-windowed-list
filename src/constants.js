// external dependencies
import noop from 'lodash/noop';

/**
 * @constant {boolean} ADD_EVENT_LISTENER_OPTIONS
 * @default
 */
export const ADD_EVENT_LISTENER_OPTIONS = !(() => {
  if (typeof window === 'undefined') {
    return false;
  }

  let hasSupport = false;

  try {
    document.createElement('div').addEventListener('test', noop, {
      get passive() {
        hasSupport = true;

        return false;
      }
    });
    /* eslint-disable no-empty */
  } catch (e) {}
  /* eslint-enable */

  return hasSupport;
})()
  ? false
  : {
    passive: true
  };

/**
 * @constant {Object} CLIENT_SIZE_KEYS
 */
export const CLIENT_SIZE_KEYS = {
  x: 'clientWidth',
  y: 'clientHeight'
};

/**
 * @constant {Object} CLIENT_START_KEYS
 */
export const CLIENT_START_KEYS = {
  x: 'clientTop',
  y: 'clientLeft'
};

/**
 * @constant {Object} INNER_SIZE_KEYS
 */
export const INNER_SIZE_KEYS = {
  x: 'innerWidth',
  y: 'innerHeight'
};

/**
 * @constant {number} MAX_CACHE_SIZE
 */
export const MAX_CACHE_SIZE = 250;

/**
 * @constant {Object} OFFSET_SIZE_KEYS
 */
export const OFFSET_SIZE_KEYS = {
  x: 'offsetWidth',
  y: 'offsetHeight'
};

/**
 * @constant {Object} OFFSET_START_KEYS
 */
export const OFFSET_START_KEYS = {
  x: 'offsetLeft',
  y: 'offsetTop'
};

/**
 * @constant {Object} OVERFLOW_KEYS
 */
export const OVERFLOW_KEYS = {
  x: 'overflowX',
  y: 'overflowY'
};

/**
 * @constant {Array<string>} OVERFLOW_VALUES
 */
export const OVERFLOW_VALUES = ['auto', 'scroll', 'overlay'];

/**
 * @constant {Object} SCROLL_SIZE_KEYS
 */
export const SCROLL_SIZE_KEYS = {
  x: 'scrollWidth',
  y: 'scrollHeight'
};

/**
 * @constant {Object} SCROLL_START_KEYS
 */
export const SCROLL_START_KEYS = {
  x: 'scrollLeft',
  y: 'scrollTop'
};

/**
 * @constant {Object} SIZE_KEYS
 */
export const SIZE_KEYS = {
  x: 'width',
  y: 'height'
};

/**
 * @constant {string} UNSTABLE_MESSAGE
 * @default
 */
export const UNSTABLE_MESSAGE = 'WindowedList failed to reach a stable state.';

/**
 * @constant {number} MAX_SYNC_UPDATES
 * @default
 */
export const MAX_SYNC_UPDATES = 100;

/**
 * @constant {Object} VALID_AXES
 */
export const VALID_AXES = {
  X: 'x',
  Y: 'y'
};

/**
 * @constant {Array<string>} VALID_AXIS_VALUES
 */
export const VALID_AXIS_VALUES = Object.keys(VALID_AXES).map((key) => {
  return VALID_AXES[key];
});

/**
 * @constant {Object} VALID_TYPES
 */
export const VALID_TYPES = {
  SIMPLE: 'simple',
  UNIFORM: 'uniform',
  VARIABLE: 'variable'
};

/**
 * @constant {Array<string>} VALID_TYPE_VALUES
 */
export const VALID_TYPE_VALUES = Object.keys(VALID_TYPES).map((key) => {
  return VALID_TYPES[key];
});

/**
 * @constant {string} DEFAULT_AXIS
 * @default
 */
export const DEFAULT_AXIS = VALID_AXES.Y;

/**
 * @constant {Object} OUTER_CONTAINER_STYLE
 */
export const OUTER_CONTAINER_STYLE = {
  height: '100%',
  maxHeight: 'inherit',
  width: '100%'
};

/**
 * @constant {Object} DEFAULT_CONTAINER_STYLE
 */
export const DEFAULT_CONTAINER_STYLE = {
  position: 'relative'
};

/**
 * @constant {Array<string>} REMEASURE_OPTIONS
 */
export const REMEASURE_OPTIONS = {
  inheritedMethods: ['getVisibleRange', 'scrollAround', 'scrollTo']
};

/**
 * @constant {Array<string>} REMEASURE_PROPERTIES
 */
export const REMEASURE_PROPERTIES = ['height', 'width'];
