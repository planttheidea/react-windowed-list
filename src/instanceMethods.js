// external dependencies
import debounce from 'lodash/debounce';
import isFunction from 'lodash/isFunction';
import isNumber from 'lodash/isNumber';
import isUndefined from 'lodash/isUndefined';
import noop from 'lodash/noop';
import raf from 'raf';
import {
  findDOMNode
} from 'react-dom';

// constants
import {
  OFFSET_SIZE_KEYS,
  OVERFLOW_KEYS,
  OVERFLOW_VALUES,
  SCROLL_START_KEYS,
  ADD_EVENT_LISTENER_OPTIONS,
  VALID_TYPES
} from './constants';

// utils
import {
  areStateValuesEqual,
  getCalculatedElementEnd,
  getCalculatedItemSizeAndItemsPerRow,
  getCalculatedSpaceBefore,
  getContainerStyle,
  getFromAndSize,
  getFromAndSizeFromListItemSize,
  getListContainerStyle,
  getOffset,
  getScrollSize,
  getViewportSize,
  hasDeterminateSize,
  setCacheSizes
} from './utils';

export const createGetContainerStyle = (instance) => {
  /**
   * @function getContainerStyle
   *
   * @description
   * create a memoized handler for getting the style object for the main container
   *
   * @returns {Object} the style object for the main container
   */
  return () => {
    const {
      axis,
      length
    } = instance.props;
    const {
      itemsPerRow
    } = instance.state;

    const bottom = Math.ceil(length / itemsPerRow) * itemsPerRow;
    const size = instance.getSpaceBefore(bottom, {});

    return getContainerStyle(axis, size);
  };
};

export const createGetItemSizeAndItemsPerRow = (instance) => {
  /**
   * @function getItemSizeAndItemsPerRow
   *
   * @description
   * get the itemSize and itemsPerRow values based on props
   *
   * @returns {{itemSize: number, itemsPerRow: number}} the itemSize and itemsPerRow
   */
  return () => {
    const {
      axis,
      useStaticSize
    } = instance.props;
    const {
      itemSize,
      itemsPerRow
    } = instance.state;

    if (useStaticSize && itemSize && itemsPerRow) {
      return {
        itemSize,
        itemsPerRow
      };
    }

    const itemElements = instance.items ? instance.items.children : [];

    return itemElements.length ? getCalculatedItemSizeAndItemsPerRow(itemElements, axis, itemSize) : {};
  };
};

export const createGetListContainerStyle = (instance) => {
  /**
   * @function getListContainerStyle
   *
   * @description
   * create a memoized handler for getting the style object for the list container
   *
   * @returns {Object} the style object for the list container
   */
  return () => {
    const {
      axis,
      usePosition,
      useTranslate3d
    } = instance.props;
    const {
      from
    } = instance.state;

    const offset = instance.getSpaceBefore(from, {});

    return getListContainerStyle(axis, usePosition, useTranslate3d, offset);
  };
};

export const createGetScrollOffset = (instance) => {
  /**
   * @function getScrollOffset
   *
   * @description
   * get the scroll offset based on props
   *
   * @returns {number} the scrollOffset to apply
   */
  return () => {
    if (!instance.outerContainer) {
      return 0;
    }

    const {
      axis
    } = instance.props;

    const scrollKey = SCROLL_START_KEYS[axis];
    const actual = instance.scrollParent === window ?
      (document.body[scrollKey] || document.documentElement[scrollKey]) :
      instance.scrollParent[scrollKey];

    const max = getScrollSize(instance.scrollParent, axis) - getViewportSize(instance.scrollParent, axis);
    const scroll = Math.max(0, Math.min(actual, max));

    return getOffset(instance.scrollParent, axis) + scroll - getOffset(instance.outerContainer, axis);
  };
};

export const createGetScrollParent = (instance) => {
  /**
   * @function getScrollParent
   *
   * @description
   * get the scroll parent element
   *
   * @returns {HTMLElement} the scroll parent
   */
  return () => {
    const {
      axis,
      scrollParentGetter
    } = instance.props;

    if (isFunction(scrollParentGetter)) {
      return scrollParentGetter();
    }

    if (!instance.outerContainer) {
      return null;
    }

    const overflowKey = OVERFLOW_KEYS[axis];

    let element = instance.outerContainer;

    while (element = element.parentElement) {
      if (~OVERFLOW_VALUES.indexOf(window.getComputedStyle(element)[overflowKey])) {
        return element;
      }
    }

    return window;
  };
};

export const createGetSizeOfListItem = (instance) => {
  /**
   * @function getSizeOfListItem
   *
   * @description
   * get the size of the list item requested
   *
   * @param {number} index the index of the list item
   * @returns {number} the size of the list item
   */
  return (index) => {
    const {
      axis,
      itemSizeEstimator,
      itemSizeGetter,
      type
    } = instance.props;
    const {
      from,
      itemSize,
      size
    } = instance.state;

    // Try the static itemSize.
    if (itemSize) {
      return itemSize;
    }

    // Try the itemSizeGetter.
    if (isFunction(itemSizeGetter)) {
      return itemSizeGetter(index);
    }

    // Try the cache.
    if (!isUndefined(instance.cache[index])) {
      return instance.cache[index];
    }

    if (instance.items) {
      const itemElements = instance.items.children;

      // Try the DOM.
      if (itemElements.length && type === VALID_TYPES.SIMPLE && index >= from && index < from + size) {
        const itemEl = itemElements[index - from];

        if (itemEl) {
          return itemEl[OFFSET_SIZE_KEYS[axis]];
        }
      }
    }

    // Try the itemSizeEstimator.
    if (isFunction(itemSizeEstimator)) {
      return itemSizeEstimator(index, instance.cache);
    }
  };
};

export const createGetSpaceBefore = (instance) => {
  /**
   * @function getSpaceBefore
   *
   * @description
   * get the space before the item requested
   *
   * @param {number} index the index of the item requested
   * @param {Object} [cache={}] the instance cache
   * @returns {number} the space before the item requested
   */
  return (index, cache = {}) => {
    if (!isUndefined(cache[index])) {
      return cache[index];
    }

    // Try the static itemSize.
    const {
      itemSize,
      itemsPerRow
    } = instance.state;

    cache[index] = itemSize ? Math.floor(index / itemsPerRow) * itemSize :
      getCalculatedSpaceBefore(cache, index, instance.getSizeOfListItem);

    return cache[index];
  };
};

export const createGetStartAndEnd = (instance) => {
  /**
   * @function getStartAndEnd
   *
   * @description
   * get the start and end values based on scroll position
   *
   * @param {number} [threshold=this.props.threshold] the pixel threshold to scroll above and below
   * @returns {{end: number, start: number}} the start and end of the window
   */
  return (threshold = instance.props.threshold) => {
    const {
      axis,
      itemSizeGetter,
      length,
      type
    } = instance.props;

    const scroll = instance.getScrollOffset();
    const start = Math.max(0, scroll - threshold);
    const calculatedEnd = scroll + getViewportSize(instance.scrollParent, axis) + threshold;
    const end = !hasDeterminateSize(type, itemSizeGetter) ? calculatedEnd :
      Math.min(calculatedEnd, instance.getSpaceBefore(length));

    return {
      end,
      start
    };
  };
};

export const createGetVisibleRange = (instance) => {
  /**
   * @function getVisibleRange
   *
   * @description
   * get the indices of the first and last items that are visible in the viewport
   *
   * @returns {Array<number>} the first and last index of the visible items
   */
  return () => {
    const {
      from,
      size
    } = instance.state;
    const {
      end,
      start
    } = instance.getStartAndEnd(0);

    const cache = {};

    let first, last, itemStart, itemEnd,
        index = from - 1,
        length = from + size;

    while (++index < length) {
      itemStart = instance.getSpaceBefore(index, cache);
      itemEnd = itemStart + instance.getSizeOfListItem(index);

      if (isUndefined(first)) {
        if (itemEnd > start) {
          first = index;
        }
      } else if (itemStart < end) {
        last = index;
      }
    }

    return [
      first,
      last
    ];
  };
};

export const createRenderItems = (instance) => {
  /**
   * @function renderItems
   *
   * @description
   * render the items that are currently visible
   *
   * @returns {ReactElement} the rendered container with the items
   */
  return () => {
    const {
      itemRenderer,
      containerRenderer
    } = instance.props;
    const {
      from,
      size
    } = instance.state;

    const items = [];

    let index = -1;

    while (++index < size) {
      items[index] = itemRenderer(from + index, index);
    }

    return containerRenderer(items, (containerRef) => {
      return instance.items = findDOMNode(containerRef);
    });
  };
};

export const createScrollAround = (instance) => {
  /**
   * @function scrollAround
   *
   * @description
   * scroll to a point that the item is within the window, but not necessarily at the top
   *
   * @param {number} index the index to scroll to in the window
   * @returns {void}
   */
  return (index) => {
    const bottom = instance.getSpaceBefore(index);
    const top = bottom - instance.getViewportSize() + instance.getSizeOfListItem(index);

    const min = Math.min(top, bottom);
    const max = Math.max(top, bottom);
    const current = instance.getScrollOffset();

    if (current <= min) {
      return instance.setScroll(min);
    }

    if (current > max) {
      return instance.setScroll(max);
    }
  };
};

export const createScrollTo = (instance) => {
  /**
   * @function scrollTo
   *
   * @description
   * scroll the element to the requested initialIndex
   *
   * @param {number} index the index to scroll to
   */
  return (index) => {
    const {
      initialIndex
    } = instance.props;

    const indexToScrollTo = isNumber(index) ? index : initialIndex;

    if (isNumber(indexToScrollTo)) {
      instance.setScroll(instance.getSpaceBefore(indexToScrollTo));
    }
  };
};

export const createSetReconcileFrameAfterUpdate = (instance) => {
  /**
   * @function setReconcileFrameAfterUpdate
   *
   * @description
   * set the frame reconciler used after componentDidUpdate
   */
  return () => {
    const {
      debounceReconciler
    } = instance.props;

    instance.reconcileFrameAfterUpdate = !isNumber(debounceReconciler) ? raf :
      debounce((updateFrame) => {
        updateFrame();
      }, debounceReconciler);
  };
};

export const createSetScroll = (instance) => {
  /**
   * @function setScroll
   *
   * @description
   * set the scroll based on the current offset
   *
   * @param {number} currentOffset the current offset
   * @returns {void}
   */
  return (currentOffset) => {
    if (!instance.scrollParent || !instance.outerContainer) {
      return;
    }

    const {
      axis
    } = instance.props;

    let offset = currentOffset + getOffset(instance.outerContainer, axis);

    if (instance.scrollParent === window) {
      return window.scrollTo(0, offset);
    }

    offset -= getOffset(instance.scrollParent, axis);

    instance.scrollParent[SCROLL_START_KEYS[axis]] = offset;
  };
};

export const createSetStateIfAppropriate = (instance) => {
  /**
   * @function setStateIfAppropriate
   *
   * @description
   * set the state if areStateValuesEqual returns true
   *
   * @param {Object} nextState the possible next state of the instance
   * @param {function} callback the callback to call once the state is set
   * @returns {void}
   */
  return (nextState, callback) => {
    return areStateValuesEqual(instance.state, nextState) ? callback() : instance.setState(nextState, callback);
  };
};

export const createUpdateFrame = (instance) => {
  /**
   * @function updateFrame
   *
   * @description
   * update the frame based on the type in props
   *
   * @param {function} callback the function to call once the frame is updated
   * @returns {void}
   */
  return (callback) => {
    const {
      type
    } = instance.props;

    instance.updateScrollParent();

    if (!isFunction(callback)) {
      callback = noop;
    }

    if (type === VALID_TYPES.UNIFORM) {
      return instance.updateUniformFrame(callback);
    }

    if (type === VALID_TYPES.VARIABLE) {
      return instance.updateVariableFrame(callback);
    }

    return instance.updateSimpleFrame(callback);
  };
};

export const createUpdateScrollParent = (instance) => {
  /**
   * @function updateScrollParent
   *
   * @description
   * update the scroll parent with the listeners it needs
   *
   * @returns {void}
   */
  return () => {
    const newScrollParent = instance.getScrollParent();

    if (newScrollParent === instance.scrollParent) {
      return;
    }

    if (instance.scrollParent) {
      instance.scrollParent.removeEventListener('scroll', instance.updateFrame);
      instance.scrollParent.removeEventListener('mousewheel', noop);
    }

    instance.scrollParent = newScrollParent;

    if (instance.scrollParent) {
      instance.scrollParent.addEventListener('scroll', instance.updateFrame, ADD_EVENT_LISTENER_OPTIONS);
      instance.scrollParent.addEventListener('mousewheel', noop, ADD_EVENT_LISTENER_OPTIONS);
    }
  };
};

export const createUpdateSimpleFrame = (instance) => {
  /**
   * @function updateSimpleFrame
   *
   * @description
   * update the frame when the type is 'simple'
   *
   * @param {function} callback the function to call once the frame is updated
   * @returns {void}
   */
  return (callback) => {
    const {
      end
    } = instance.getStartAndEnd();

    if (!instance.items || getCalculatedElementEnd(instance.items.children, instance.props) > end) {
      return callback();
    }

    const {
      pageSize,
      length
    } = instance.props;

    const size = Math.min(instance.state.size + pageSize, length);

    instance.setStateIfAppropriate({
      size
    }, callback);
  };
};

export const createUpdateUniformFrame = (instance) => {
  /**
   * @function updateUniformFrame
   *
   * @description
   * update the frame when the type is 'uniform'
   *
   * @param {function} callback the function to call once the frame is updated
   * @returns {void}
   */
  return (callback) => {
    const itemSizeAndItemsPerRow = instance.getItemSizeAndItemsPerRow();
    const {
      itemSize,
      itemsPerRow
    } = itemSizeAndItemsPerRow;

    if (!itemSize || !itemsPerRow) {
      return callback();
    }

    const {
      start,
      end
    } = instance.getStartAndEnd();

    const calculatedFrom = Math.floor(start / itemSize) * itemsPerRow;
    const calulatedSize = (Math.ceil((end - start) / itemSize) + 1) * itemsPerRow;
    const fromAndSize = getFromAndSize(calculatedFrom, calulatedSize, itemsPerRow, instance.props);

    return instance.setStateIfAppropriate({
      ...fromAndSize,
      ...itemSizeAndItemsPerRow
    }, callback);
  };
};

export const createUpdateVariableFrame = (instance) => {
  /**
   * @function updateVariableFrame
   *
   * @description
   * update the frame when the type is 'variable'
   *
   * @param {function} callback the function to call once the frame is updated
   * @returns {void}
   */
  return (callback) => {
    if (!instance.items) {
      return;
    }

    const {
      axis,
      itemSizeGetter
    } = instance.props;
    const {
      from: currentFrom,
      size: currentSize
    } = instance.state;

    if (!itemSizeGetter) {
      setCacheSizes(currentFrom, instance.items, axis, instance.cache);
    }

    const currentFromAndSize = {
      from: currentFrom,
      size: currentSize
    };

    const fromAndSize = getFromAndSizeFromListItemSize(instance.getStartAndEnd(), instance.props,
      instance.getSizeOfListItem, currentFromAndSize);

    instance.setStateIfAppropriate(fromAndSize, callback);
  };
};
