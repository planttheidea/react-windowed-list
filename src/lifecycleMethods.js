// external dependencies
import noop from 'lodash/noop';
import raf from 'raf';
import {
  findDOMNode
} from 'react-dom';

// constants
import {
  MAX_SYNC_UPDATES,
  ADD_EVENT_LISTENER_OPTIONS,
  UNSTABLE_MESSAGE
} from './constants';

// utils
import {
  getFromAndSize
} from './utils';

/**
 * @function getInitialState
 *
 * @description
 * get the initial state of the instance
 *
 * @returns {{from: number, itemsPerRow: number, size: number}} the initial state
 */
export const getInitialState = () => {
  return {
    from: 0,
    itemsPerRow: 0,
    size: 0
  };
};

export const createComponentDidMount = (instance) => {
  /**
   * @function componentDidMount
   *
   * @description
   * on mount, update the frame with the desired scroll position
   */
  return () => {
    instance.outerContainer = findDOMNode(instance);

    if (!instance.props.isHidden) {
      raf(() => {
        instance.updateFrame(instance.scrollTo);
      });
    }
  };
};

export const createComponentDidUpdate = (instance) => {
  /**
   * @function componentDidUpdate
   *
   * @description
   * update the frame position, cutting off after a certain point if unstable
   *
   * @returns {void}
   */
  return () => {
    if (instance.unstable) {
      return;
    }

    if (++instance.updateCounter > MAX_SYNC_UPDATES) {
      instance.unstable = true;

      /* eslint-disable no-console */
      return console.error(UNSTABLE_MESSAGE);
      /* eslint-enable */
    }

    if (!instance.updateCounterTimeoutId) {
      instance.updateCounterTimeoutId = raf(() => {
        instance.updateCounter = 0;

        delete instance.updateCounterTimeoutId;
      });
    }

    if (!instance.props.isHidden) {
      instance.reconcileFrameAfterUpdate(instance.updateFrame);
    }
  };
};

export const createComponentWillMount = (instance) => {
  /**
   * @function componentWillMount
   *
   * @description
   * prior to mount, set the initial state
   */
  return () => {
    const {
      initialIndex
    } = instance.props;
    const itemsPerRow = 1;
    const fromAndSize = getFromAndSize(initialIndex, 0, itemsPerRow, instance.props);

    instance.setReconcileFrameAfterUpdate();

    instance.setState({
      ...fromAndSize,
      itemsPerRow
    });
  };
};

export const createComponentWillReceiveProps = (instance) => {
  /**
   * @function componentWillReceiveProps
   *
   * @description
   * when props are received, set the state if the next calculated state is different
   *
   * @param {Object} nextProps the incoming props
   */
  return (nextProps) => {
    const {
      debounceReconciler
    } = instance.props;
    const {
      from,
      itemsPerRow,
      size
    } = instance.state;

    if (nextProps.debounceReconciler !== debounceReconciler) {
      instance.setReconcileFrameAfterUpdate();
    }

    instance.setStateIfAppropriate(getFromAndSize(from, size, itemsPerRow, nextProps), noop);
  };
};

export const createComponentWillUnmount = (instance) => {
  /**
   * @function componentWillUnmount
   *
   * @description
   * before unmount, remove any listeners applied to the scroll container
   */
  return () => {
    if (instance.scrollParent) {
      instance.scrollParent.removeEventListener('scroll', instance.updateFrame, ADD_EVENT_LISTENER_OPTIONS);
      instance.scrollParent.removeEventListener('mousewheel', noop, ADD_EVENT_LISTENER_OPTIONS);
    }

    instance.outerContainer = null;
  };
};
