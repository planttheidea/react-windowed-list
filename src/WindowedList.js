// external dependencies
import React, {Component} from 'react';

// classes
import WindowedListRenderer from './WindowedListRenderer';

export const createGetVisibleRange = (instance) => {
  /**
   * @function getVisibleRange
   *
   * @description
   * get the visible range on the windowed list component
   *
   * @returns {Array<number>}
   */
  return () => {
    if (!instance.ref || !instance.ref.originalComponent) {
      return [0, 0];
    }

    return instance.ref.originalComponent.getVisibleRange();
  };
};

export const createScrollAround = (instance) => {
  /**
   * @function scrollAround
   *
   * @description
   * scroll around the index passed on the windowed list component
   *
   * @param {number} index the index to scroll around
   * @returns {Array<number>}
   */
  return (index) => {
    if (!instance.ref || !instance.ref.originalComponent) {
      return;
    }

    return instance.ref.originalComponent.scrollAround(index);
  };
};

export const createScrollTo = (instance) => {
  /**
   * @function scrollTo
   *
   * @description
   * scroll to the index passed on the windowed list component
   *
   * @param {number} index the index to scroll to
   * @returns {Array<number>}
   */
  return (index) => {
    if (!instance.ref || !instance.ref.originalComponent) {
      return;
    }

    return instance.ref.originalComponent.scrollTo(index);
  };
};

export const createSetRef = (instance) => {
  return (element) => {
    instance.ref = element;
  };
};

class WindowedList extends Component {
  static displayName = 'WindowedList';

  // instance values
  ref = null;

  // instance methods
  getVisibleRange = createGetVisibleRange(this);
  scrollAround = createScrollAround(this);
  scrollTo = createScrollTo(this);
  setRef = createSetRef(this);

  render() {
    return (
      /* eslint-disable prettier */
      <WindowedListRenderer
        {...this.props}
        ref={this.setRef}
      />
      /* eslint-enable */
    );
  }
}

export default WindowedList;
