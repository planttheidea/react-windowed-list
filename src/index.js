// components
import WindowedList from './WindowedList';

export default WindowedList;
