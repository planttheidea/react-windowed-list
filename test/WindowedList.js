// test
import test from 'ava';
import React from 'react';
import sinon from 'sinon';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

// src
import * as component from 'src/WindowedList';

const WindowedList = component.default;

test('if getVisibleRange will execute the same method on the original component', (t) => {
  const visibleRange = [0, 10];

  const instance = {
    ref: {
      originalComponent: {
        getVisibleRange: sinon.stub().returns(visibleRange)
      }
    }
  };

  const getVisibleRange = component.createGetVisibleRange(instance);

  const result = getVisibleRange();

  t.true(instance.ref.originalComponent.getVisibleRange.calledOnce);

  t.is(result, visibleRange);
});

test('if getVisibleRange will return a default when originalComponent does not exist on the ref', (t) => {
  const instance = {
    ref: {}
  };

  const getVisibleRange = component.createGetVisibleRange(instance);

  const result = getVisibleRange();

  t.deepEqual(result, [0, 0]);
});

test('if getVisibleRange will return a default when the ref does not exist', (t) => {
  const instance = {};

  const getVisibleRange = component.createGetVisibleRange(instance);

  const result = getVisibleRange();

  t.deepEqual(result, [0, 0]);
});

test('if scrollAround will execute the same method on the original component', (t) => {
  const index = 10;

  const instance = {
    ref: {
      originalComponent: {
        scrollAround: sinon.spy()
      }
    }
  };

  const scrollAround = component.createScrollAround(instance);

  scrollAround(index);

  t.true(instance.ref.originalComponent.scrollAround.calledOnce);
  t.true(instance.ref.originalComponent.scrollAround.calledWith(index));
});

test('if scrollAround will not throw when there is no original component', (t) => {
  const index = 10;

  const instance = {
    ref: {}
  };

  const scrollAround = component.createScrollAround(instance);

  t.notThrows(() => {
    scrollAround(index);
  });
});

test('if scrollAround will not throw when there is no ref', (t) => {
  const index = 10;

  const instance = {};

  const scrollAround = component.createScrollAround(instance);

  t.notThrows(() => {
    scrollAround(index);
  });
});

test('if scrollTo will execute the same method on the original component', (t) => {
  const index = 10;

  const instance = {
    ref: {
      originalComponent: {
        scrollTo: sinon.spy()
      }
    }
  };

  const scrollTo = component.createScrollTo(instance);

  scrollTo(index);

  t.true(instance.ref.originalComponent.scrollTo.calledOnce);
  t.true(instance.ref.originalComponent.scrollTo.calledWith(index));
});

test('if scrollTo will not throw when there is no original component', (t) => {
  const index = 10;

  const instance = {
    ref: {}
  };

  const scrollTo = component.createScrollTo(instance);

  t.notThrows(() => {
    scrollTo(index);
  });
});

test('if scrollTo will not throw when there is no ref', (t) => {
  const index = 10;

  const instance = {};

  const scrollTo = component.createScrollTo(instance);

  t.notThrows(() => {
    scrollTo(index);
  });
});

test('if WindowedList renders correctly with default props', (t) => {
  const props = {};

  const wrapper = shallow(<WindowedList {...props} />);

  t.snapshot(toJson(wrapper));
});
