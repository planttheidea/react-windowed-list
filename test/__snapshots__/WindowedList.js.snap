// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if WindowedList renders correctly when type is not simple 1`] = `
<div
  style={
    Object {
      "height": 500,
    }
  }
>
  <WindowedList
    length={1000}
    type="uniform"
  >
    <WindowedList
      axis="y"
      containerRenderer={[Function]}
      initialIndex={0}
      isHidden={false}
      isLazy={false}
      itemRenderer={[Function]}
      length={1000}
      minSize={1}
      pageSize={10}
      size={
        Object {
          "height": 0,
          "width": 0,
        }
      }
      threshold={100}
      type="uniform"
      usePosition={false}
      useStaticSize={false}
      useTranslate3d={false}
    >
      <div
        style={
          Object {
            "height": "100%",
            "maxHeight": "inherit",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <div
            style={
              Object {
                "WebkitTransform": "translate(0px, 0px)",
                "msTransform": "translate(0px, 0px)",
                "transform": "translate(0px, 0px)",
              }
            }
          >
            <div>
              <div>
                0
              </div>
              <div>
                1
              </div>
              <div>
                2
              </div>
              <div>
                3
              </div>
              <div>
                4
              </div>
              <div>
                5
              </div>
              <div>
                6
              </div>
              <div>
                7
              </div>
              <div>
                8
              </div>
              <div>
                9
              </div>
            </div>
          </div>
        </div>
      </div>
    </WindowedList>
  </WindowedList>
</div>
`;

exports[`if WindowedList renders correctly with default props 1`] = `
<div
  style={
    Object {
      "height": 500,
    }
  }
>
  <WindowedList
    length={1000}
  />
</div>
`;
